hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/message/MessageState.java

```diff
@@ -1,33 +1,53 @@
 package pl.allegro.tech.hermes.frontend.publishing.message;
 
-import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.SENDING_TO_KAFKA;
-import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.SENT_TO_KAFKA;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.READING;
 
 public class MessageState {
 
     public enum State {
-        WAITING_FOR_FIRST_PORTION_OF_DATA,
-        PARSING,
-        PARSED,
+        READING,
+        READING_TIMEOUT,
+        FULLY_READ,
         SENDING_TO_KAFKA_PRODUCER_QUEUE,
+        ERROR_IN_SENDING_TO_KAFKA_PRODUCER_QUEUE,
         SENDING_TO_KAFKA,
-        SENT_TO_KAFKA
+        SENT_TO_KAFKA,
+        DELAYED_SENDING,
+        DELAYED_PROCESSING
     }
 
-    private volatile State state = State.WAITING_FOR_FIRST_PORTION_OF_DATA;
+    private final String messageId;
 
     public MessageState() {
+        messageId = MessageIdGenerator.generate();
+    }
+
+    private volatile boolean delayed = false;
+    private AtomicReference<State> state = new AtomicReference<>(READING);
+
+    public boolean compareAndSet(State expect, State update) {
+        return state.compareAndSet(expect, update);
+    }
+
+    public void set(State state) {
+        this.state.set(state);
+    }
+
+    public State get() {
+        return state.get();
     }
 
-    public synchronized void setState(State state) {
-        this.state = state;
+    public boolean isDelayed() {
+        return delayed;
```

Comment:
isDelayed() name can be misleading, I changed it to isDelayedProcessing(). As for states it is a tricky part. What is important is that when timeout occurs and request is on processing message state (for example converting it from json -> avro or validating it) then 408 is not returned (as it was in previous version of frontend) but response is delayed until message arrives to kafka producer buffer, and after that 202 is returned immediately. This can break 65ms barrier but imo is better than retrying, especially bigger messages.